### PR TITLE
Adding basic pre-commit hooks

### PR DESCRIPTION
**Pull request by jpodivin**
Elementary pre-commit hooks were added to guard against
some common style issues, such as trailing whitespace.

Use of pre-commit should prevent style violations from creeping
in without need for any CI resources. As all the checks are run
on the developers own machine.

Files were edited in order to comply with style guidelines.

Signed-off-by: Jiri Podivin <jpodivin@gmail.com>

### CODE REVIEW

**Comment by haifeng-jin** [comment]
I think this is something nice to have if we can have it setup in codespaces, can we? @jpodivin Otherwise, we don't have a good way to encourage the user to use them.

@qlzh727 Do you think the [large file checker](https://github.com/keras-team/keras/pull/16351/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R10) will work with our codebase in the future? 

Note: This pull request is adding a new file, which won't be properly copied by copybara.

**Comment by qlzh727** [comment]
Large file check should be fine. We don't intend to have any data file checked into our repo, so a check for that is ok.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+    -   id: trailing-whitespace
+    -   id: end-of-file-fixer
+    -   id: check-yaml
```
Comment (fchollet) [inline .pre-commit-config.yaml:9]: We don't have any yaml files, do we?

**Comment by haifeng-jin** [comment]
The big concern is if we do not enforce this pre-commit for internal contributors, it would break it.
All external contributor who setup this pre-commit would need to fix it everytime that an internal commit that breaks it.

**Comment by jpodivin** [comment]
Following the discussion from April I've taken look into how would pre-commit work with the codespaces.
Unfortunately I'm not using them myself, as I have only a personal account. But working with their docs I've come to some conclusions.

The integration shouldn't be a problem. Pre-commit scripts can be easily installed in basic venv, VM or container, with `pip` being the most straightforward way to manage the installation, although `brew` and `conda` are also supported. 

I've verified that pre-commit works with the provided dev container file, and I'll be submitting a new PR once it's finished.

**Comment by haifeng-jin** [comment]
I don't think the internal dev infra can support such level of precommit.
The only precommit we can merge is a `.git/hook` file that user can use locally to run `sh shell/lint.sh` at every commit.

Otherwise, internal developers can never run these hooks and will keep breaking the external checks.

**Comment by jpodivin** [comment]
I was under impression that the internal development is running in containerized environment. In that case it's relatively simple matter to integrate pre-commit as a dependency and initialize it after container is created. If users can run arbitrary scripts, like `lint.sh`, it really shouldn't be a problem.

**Comment by haifeng-jin** [comment]
The value of precommit is that it runs automatically. I do not see an easy way to do this in our dev infra. The `lint.sh` is manually run by developers.

**Comment by jpodivin** [comment]
I'm not sure I understand why it wouldn't run automatically.

Pre-commit run is normally initialized by commit action, assuming the hooks are installed. Which they would be, provided the `devcontainer.json` is adjusted accordingly, as it is in my PR.

**Pull request closed without merging**